runorders.cpp

```diff
@@ -550,71 +569,71 @@ void Game::Do1Destroy(ARegion *r, Object *o, Unit *u) {
 		return;
 	}
 
-	if (o->CanModify()) {
-		int maxStructurePoints;	// fully built structure points (equal to cost)
-		int structurePoints;	// current structure points
-		int destroyablePoints;	// how much points can be destroyed per turn for this structure type
-		int destroyPower;		// how much unit can destroy during this turn
-		int willDestroy;		// how muh points will be destroyed this turn
+	if (!o->CanModify()) {
+		u->Error("DESTROY: Can't destroy that.");
+		forlist(&o->units) {
+			((Unit *) elem)->destroy = 0;
+		}
 
-		maxStructurePoints = ObjectDefs[o->type].cost;
-		structurePoints = maxStructurePoints - o->incomplete;
+		return;
+	}
 
-		if (Globals->DESTROY_BEHAVIOR == DestroyBehavior::INSTANT) {
-			destroyablePoints = structurePoints;
-			destroyPower = structurePoints;
-		}
-		else {
-			destroyablePoints = std::max(Globals->MIN_DESTROY_POINTS, maxStructurePoints * Globals->MAX_DESTROY_PERCENT / 100);
-			destroyablePoints = std::max(0, std::min(structurePoints, destroyablePoints) - o->destroyed);
+	// ships can be destroyed in instant
+	DestroyBehavior behavior = o->IsFleet()
+		? DestroyBehavior::INSTANT
+		: Globals->DESTROY_BEHAVIOR;
 
-			if (Globals->DESTROY_BEHAVIOR == DestroyBehavior::PER_SKILL) {
-				destroyPower = std::max(1, u->GetSkill(S_BUILDING)) * u->GetMen();
-			}
-			else {
-				destroyPower = u->GetMen();
-			}
-		}
+	int structurePoints = behavior == DestroyBehavior::INSTANT
+		? get_structure_points(o, 0, 100)
+		: get_structure_points(o, Globals->MIN_DESTROY_POINTS, Globals->MAX_DESTROY_PERCENT);
 
-		willDestroy = std::min(destroyablePoints, destroyPower);
-		if (willDestroy == 0) {
-			u->Error(AString("Can't destroy ") + *(o->name) + " more.");
-			forlist(&o->units) {
-				((Unit *) elem)->destroy = 0;
-			}
+	int destroyPower;
+	switch (behavior) {
+		case DestroyBehavior::PER_SKILL:
+			destroyPower = u->GetMen() * std::max(1, u->GetSkill(S_BUILDING));
+			break;
 
-			return;
-		}
+		case DestroyBehavior::PER_FIGURE:
+			destroyPower = u->GetMen();
+			break;
 
-		int remainingSP = structurePoints - willDestroy;
-		if (remainingSP > 0) {
-			o->destroyed += willDestroy;
-			o->incomplete += willDestroy;
+		default:
+			destroyPower = structurePoints;
+	}
 
-			u->Event(AString("Destroys ") + willDestroy + " structure points from the " + *(o->name) + ".");
+	int willDestroy = std::min(structurePoints, destroyPower);
+	if (willDestroy == 0) {
+		u->Error(AString("Can't destroy ") + *(o->name) + " more.");
```

Comment:
```suggestion
		u->error("Can't destroy " + string(o->name->const_str()) + " more.");
```

this is the new way :)